### ISSUE TITLE
Visual Studio 2017 + x64 overall exceptions when compiling in Release (default)and also Debug mode

### ISSUE THREAD

**Issue by rafalsk** [randombit/botan#1167]
I haven't compiled with debug symbols but this should be easy to reproduce.
```
`Botan::AutoSeeded_RNG rng;
Botan::secure_vector<uint8_t> privKeyDER;
Botan::Curve25519_PrivateKey privKey = Botan::Curve25519_PrivateKey(rng);
privKeyDER = privKey.private_key_bits();`
```
1) throws "vector<T> too long" at the last line

2) the privKey.check_key() returns false, what am I doing wrong?

**Comment by randombit** [comment]
Can you post a complete source code along with the output of `Botan::version_string()` (in `version.h`)? I can't replicate either issue with master.

**Comment by rafalsk** [comment]
@randombit 
The complete source code is as follows:
`// BotanTest.cpp : Defines the entry point for the console application.
//
#include "stdafx.h"
#include "botan_all.h"

int main()
{
        std::cout <<  Botan::version_string(); //throws Access Violation exception
	Botan::AutoSeeded_RNG rng;
	Botan::secure_vector<uint8_t> privKeyDER;
	Botan::Curve25519_PrivateKey privKey = Botan::Curve25519_PrivateKey(rng);
	privKeyDER = privKey.private_key_bits();
    return 0;
}
I have created a separate new project to reproduce the issue.
I am using Visual Studio 2017, running on Windows 10 Pro, compile mode is set to x64.

**Comment by randombit** [comment]
`std::cout <<  Botan::version_string(); //throws Access Violation exception`

Well something is definitely wrong then. since this function is just returning a constant string. I'm not sure what the problem might be though, this kind of feels like some Windows DLL issue but nothing jumping out in my mind what it might be.

Do the tests pass?

**Comment by rafalsk** [comment]
**Update**
Downloaded from 
https://codeload.github.com/randombit/botan/zip/master

1) in command line:
python configure.py --cc=msvc --amalgamation 
2)from x64 VS command prompt : nmake
3) copied boran.dll to the folder of executable
4) added botan.lib to be included by linker
5) copied and added the genrated cpp and h files , added them to the project
6) compiled fine and throws

I'll uploaded the project to a Test Server and report back

**Comment by rafalsk** [comment]
exactly same thing happens on our Test Server
Configuration:
Windows Server 2012 R2 (latest updates), Visual Studio 2016 Ent.

**Comment by rafalsk** [comment]
you can download the entire project with compiled libraries from
gridnet.org/botantest.zip

**Comment by rafalsk** [comment]
Update: 
I still haven't resolved this issue. WIll try without the amalgemation option now.
Another bug:
Even when using amalgemation, the resulting files fail to include pkcs11t.h, pkcs11.h and pkcs11f.h

**Comment by rafalsk** [comment]
can you confirm that you were able to reproduce the issue ? Just so I feel calm its being worked upon. Thanks.

**Comment by randombit** [comment]
Sorry I just tried reproing with the source you linked to and everything works fine for me on Linux with gcc. I don't have any Windows machines.

The fact that it works on debug mode but not with optimizations kind of suggests a compiler bug to me though that's not certain. Can you clarify which version of VC you are using, "Visual Studio 2016" does not seem to exist that I can find, just 2015 and 2017 (both of which are tested in CI without problems). The version reported by running `cl` on the command line might help clear that up.

Re `botan/build.h` there should be a symlink (or I guess on Windows a copy) of this file into `build/include/botan`, strange that it is not there...

**Comment by rafalsk** [comment]
@randombit , what I meant was visual studio 2017.  The output of CL:
`Microsoft (R) C/C++ Optimizing Compiler Version 19.11.25507.1 for x64   `

**Comment by rafalsk** [comment]
@randombit , You said this should work on Visual Studio 2015, so I've spent some time figuring out how to apporach the subject. VS 2015 is not available for sale anymore, but VS 2017 does support Platform Toolkits and compilers ranging back to VS 2010.

So, I've installed Platform Toolkit 14.0 (Visual Studio 2015) and compiled Botan with the very same compiler and librararies. The output of CL is:
`Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24225.1 for x64`
and still I'm getting the very same exceptions.
